services/faucet/src/modules/faucet/faucet.service.ts

```diff
@@ -0,0 +1,91 @@
+import { ethereum, general } from '@faucet/config';
+import { Inject, Injectable } from '@nestjs/common';
+import { ConfigType } from '@nestjs/config';
+import { InjectRepository } from '@nestjs/typeorm';
+import { ethers } from 'ethers';
+import { Repository } from 'typeorm';
+import {
+  BlockchainService,
+  ContractTypes,
+} from '../blockchain/blockchain.service';
+import { CreateRequestDto } from './dto';
+import { RequestEntity } from './entity';
+import RequestLimit from './types/RequestLimit.enum';
+import { filterRequestsOfThisWeek } from './utils';
+
+@Injectable()
+export class FaucetService {
+  privateKey: Buffer;
+  constructor(
+    @InjectRepository(RequestEntity)
+    private readonly requestRepository: Repository<RequestEntity>,
+    @Inject(general.KEY)
+    private readonly generalConfig: ConfigType<typeof general>,
+    @Inject(ethereum.KEY)
+    private readonly ethereumConfig: ConfigType<typeof ethereum>,
+    private readonly blockchainService: BlockchainService
+  ) {
+    this.privateKey = Buffer.from(
+      this.ethereumConfig.privateSigningKey!,
+      'hex'
+    );
+  }
+  public async registerRequest(
+    requestDto: CreateRequestDto
+  ): Promise<RequestEntity> {
+    if (!(requestDto.amount in RequestLimit))
+      throw new Error(
+        `You must ask for either ${RequestLimit.ONE}; ${RequestLimit.FIVE}; ${RequestLimit.TEN} or ${RequestLimit.FIFTY}`
+      );
+    requestDto.address = ethers.utils.getAddress(requestDto.address);
+    const requestsForAddress = await this.requestRepository.find({
+      address: requestDto.address,
+    });
+
+    const requestsOfThisWeek = filterRequestsOfThisWeek(requestsForAddress);
+    if (requestsOfThisWeek && requestsOfThisWeek.length > 0) {
+      const total = requestsOfThisWeek
+        .map((r) => r.amount)
+        .reduce((curr, next) => curr + next);
+      if (total >= +this.generalConfig.maxRewardPerWeek!)
+        throw new Error(
```

Comment:
Wouldn't it be better to return more specific errors from Nestjs, or at least InternalServerErrorException?

Reply:
yeah, I thought about creating some custom errors also, but this is general in all services we have for now, we just use Error. Internal could be better here, yes